### PR TITLE
Use badge caps for price bumps

### LINKED ISSUE #1906: As an event that might sell out, I want to raise the badge price after a certain # of badges sold
If I don't sell to the badge limit specified, I want the badge price to increase after certain dates like it currently does. So, for example, I want to say that "the badge price should increase to $50 either after 5000 badges have been sold or after September 1st."

The slightly tricky thing here is that we never want a badge price to _lower_ again, but (unlike dates) the 'number of badges sold' is something that can change.

The more 'fool-proof' way to do this is to store the badge price in the database, and only update it if the update is an increase. However, we can alternatively use the "Refunded" status as part of the 'badges sold' calculation, and trust that admins will actually correctly use the system and not just delete badges willy-nilly. I don't really want to store the badge price in the database, but... admins and data integrity don't always mix.

### PR DESCRIPTION
**Pull request by kitsuta**
Allows admins to set a badge cap for each price bump, rather than just relying on dates. Fixes #1906.

To add a badge-cap-based price bump, simply add it in the same place in the config as the dates - the system automatically sorts the type of price bump it is based on whether it can be cast to a Date type or not.

This is the simple version, which doesn't store the badge price permanently; thus, there is a chance that the badge price can go _down_ if, e.g., several badges are deleted. However, I'm leaning towards that being preferable, as changing a single config option (or adding fake badges) to force a badge price to stay high is easier than trying to modify the database if the badge price was erroneously bumped (like if a group of 50 people all registered and then requested a refund or something silly).

### CODE REVIEW

**Comment by kitsuta** [comment]
One thing this PR doesn't handle is communicating to attendees how many badges there are left at a certain price. I figured most events wouldn't want to put that out there, since the number may fluctuate (and if it's going slow, that could make the event look silly). Since `BADGES_SOLD` is accessible via json, interested events can create their own pages for attendees to keep track if they so desire. I'd be interested in hearing others' opinions on this, though.

**Comment by kitsuta** [comment]
Oh, on an unrelated note, I attempted to make unit tests for this but couldn't figure out how to reload the Config object to recalcuate `PRICE_BUMPS` and `PRICE_LIMITS` after monkeypatching different settings in.

File: uber/config.py [hunk 1]
```diff
@@ -345,8 +348,14 @@ def _unrepr(d):
...
 c.PRICE_BUMPS = {}
+c.PRICE_LIMITS = {}
 for _opt, _val in c.BADGE_PRICES['attendee'].items():
-    c.PRICE_BUMPS[c.EVENT_TIMEZONE.localize(datetime.strptime(_opt, '%Y-%m-%d'))] = _val
+    try:
+        date = c.EVENT_TIMEZONE.localize(datetime.strptime(_opt, '%Y-%m-%d'))
+    except ValueError:
+        c.PRICE_LIMITS[int(_opt)] = _val
+    else:
+        c.PRICE_BUMPS[date] = _val
```
Comment (EliAndrewC) [inline uber/config.py]: Interesting.  So is the idea that the config file will now look something like

```
[badge_prices]
[[attendee]]

# Raise to $55 on Sept 1 or after the first 1000 badges are sold
2016-09-01 = 55
1000 = 55

# Raise to $60 on Oct 1 or after the first 2000 badges are sold
2016-09-01 = 60
2000 = 60

# etc
```

That seems sensible.  We should probably add some extra comments to ``configspec.ini`` to explain this.

**Comment by EliAndrewC** [comment]
Good stuff, I like how this is put together, looks great.

> I attempted to make unit tests for this but couldn't figure out how to reload the Config object to recalculate PRICE_BUMPS and PRICE_LIMITS after monkeypatching different settings in.

Since ``c.PRICE_BUMPS`` and ``c.PRICE_LIMITS`` are the only relevant options, I would have expected you to be able to just say ``monkeypatch.setattr(c, 'PRICE_BUMPS', {...})`` and ``monkeypatch.setattr(c, 'PRICE_LIMITS', {...})``.  That would be the most straightforward approach and should work fine in this case.

File: uber/config.py (hunk 1 above)
Comment (kitsuta) [inline uber/config.py]: Good idea!

**Comment by kitsuta** [comment]
Hm, I guess I was thinking that part of the unit tests were testing how the config was loaded in... but that's not very unit-ty. I'll go ahead and do that.

**Comment by kitsuta** [comment]
OK, I've added the unit tests, which are slightly incomplete because I couldn't figure out how to test paid_by_group badges.

Also, is there a better way to do these unit tests rather than relying on the actual session?

**Comment by EliAndrewC** [comment]
+1

> Also, is there a better way to do these unit tests rather than relying on the actual session?

The ``c.get_attendee_price`` method you're testing never references the session directly and instead makes use of ``c.BADGES_SOLD``, which itself does a query.  So the most theoretically correct way to do this would be to have two sets of tests:
-  One set of tests for ``c.BADGES_SOLD`` which would be basically what you have now, where you insert badges with different statuses and then make assertions about whether or not they affect the count.  (This is where tests about group badges would come in.)
- A second sets of tests for ``c.get_attendee_price`` that uses ``monkeypatch`` to simply set different dates and different values for ``c.BADGES_SOLD`` and make assertions about how those counts/dates affect the final price.

With that being said, what you have looks fairly solid overall, and I don't think we need to hold up merging this just to flesh out the unit tests a little.

**Comment by kitsuta** [comment]
Yeah, I actually tried what you said, but it turns out that `monkeypatch` can't mock values for @property functions by default. I figured it was better to keep it simple for now rather than try to mess with the kinds of solutions I was finding on Stack Overflow and Reddit.

**Comment by EliAndrewC** [comment]
> Yeah, I actually tried what you said, but it turns out that monkeypatch can't mock values for @property functions by default.

Ahh, for future reference the trick is to mock it at the class level instead of the instance level.  In other words:

```python
from uber.config import c, Config
monkeypatch.setattr(c, 'BADGES_SOLD', 100)  # doesn't work
monkeypatch.setattr(Config, 'BADGES_SOLD', 100)  # works
```

That distinction isn't immediately obvious, but it's because of the semantics of attribute assignment in Python.

**Comment by kitsuta** [comment]
Oops, we'll have to keep that in mind for later.

**Pull request merged**